From 840f834c1462f4597ceebf76b4fc1ed06de4a73e Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 6 Jul 2013 23:05:19 -0400
Subject: [PATCH] Anti Dupe

Remove CraftBukkit "Feature" of ItemStack's having infinite use if less than 0 stack size.
Blocks most cases of < 0 itemstack dupe bugs
---
 src/main/java/net/minecraft/server/DispenseBehaviorItem.java  | 1 +
 src/main/java/net/minecraft/server/EntityAgeable.java         | 2 +-
 src/main/java/net/minecraft/server/EntityHorse.java           | 2 +-
 src/main/java/net/minecraft/server/EntityHuman.java           | 2 +-
 src/main/java/net/minecraft/server/ItemStack.java             | 1 +
 src/main/java/net/minecraft/server/PlayerConnection.java      | 2 +-
 src/main/java/net/minecraft/server/PlayerInteractManager.java | 2 +-
 src/main/java/net/minecraft/server/PlayerInventory.java       | 2 +-
 src/main/java/net/minecraft/server/Slot.java                  | 1 +
 src/main/java/net/minecraft/server/TileEntityDispenser.java   | 1 +
 10 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/src/main/java/net/minecraft/server/DispenseBehaviorItem.java b/src/main/java/net/minecraft/server/DispenseBehaviorItem.java
index 14ca919..e9adf3c 100644
--- a/src/main/java/net/minecraft/server/DispenseBehaviorItem.java
+++ b/src/main/java/net/minecraft/server/DispenseBehaviorItem.java
@@ -10,6 +10,7 @@ public class DispenseBehaviorItem implements IDispenseBehavior {
     public DispenseBehaviorItem() {}
 
     public final ItemStack a(ISourceBlock isourceblock, ItemStack itemstack) {
+        if (itemstack != null && itemstack.count < 0) itemstack.count = 0; // EMC
         ItemStack itemstack1 = this.b(isourceblock, itemstack);
 
         this.a(isourceblock);
diff --git a/src/main/java/net/minecraft/server/EntityAgeable.java b/src/main/java/net/minecraft/server/EntityAgeable.java
index 7ddca48..11ab192 100644
--- a/src/main/java/net/minecraft/server/EntityAgeable.java
+++ b/src/main/java/net/minecraft/server/EntityAgeable.java
@@ -57,7 +57,7 @@ public abstract class EntityAgeable extends EntityCreature {
 
                         if (!entityhuman.abilities.canInstantlyBuild) {
                             --itemstack.count;
-                            if (itemstack.count == 0) {  // CraftBukkit - allow less than 0 stacks as "infinite"
+                            if (itemstack.count <= 0) { // EMC
                                 entityhuman.inventory.setItem(entityhuman.inventory.itemInHandIndex, (ItemStack) null);
                             }
                         }
diff --git a/src/main/java/net/minecraft/server/EntityHorse.java b/src/main/java/net/minecraft/server/EntityHorse.java
index d9f2f3e..8ac71f9 100644
--- a/src/main/java/net/minecraft/server/EntityHorse.java
+++ b/src/main/java/net/minecraft/server/EntityHorse.java
@@ -642,7 +642,7 @@ public class EntityHorse extends EntityAnimal implements IInventoryListener {
                 }
 
                 if (flag) {
-                    if (!entityhuman.abilities.canInstantlyBuild && --itemstack.count == 0) {
+                    if (!entityhuman.abilities.canInstantlyBuild && --itemstack.count <= 0) { // EMC
                         entityhuman.inventory.setItem(entityhuman.inventory.itemInHandIndex, (ItemStack) null);
                     }
 
diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index c58b21b..fffc464 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -529,7 +529,7 @@ public abstract class EntityHuman extends EntityLiving implements ICommandListen
     public EntityItem a(ItemStack itemstack, boolean flag, boolean flag1) {
         if (itemstack == null) {
             return null;
-        } else if (itemstack.count == 0) {
+        } else if (itemstack.count <= 0) { // EMC
             return null;
         } else {
             EntityItem entityitem = new EntityItem(this.world, this.locX, this.locY - 0.30000001192092896D + (double) this.getHeadHeight(), this.locZ, itemstack);
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 3699991..dd3a580 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -54,6 +54,7 @@ public final class ItemStack {
         ItemStack itemstack = new ItemStack();
 
         itemstack.c(nbttagcompound);
+        if (itemstack.count < 0) itemstack.count = 0; // EMC
         return itemstack.getItem() != null ? itemstack : null;
     }
 
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 91de23b..e1bc778 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -664,7 +664,7 @@ public class PlayerConnection implements PacketPlayInListener {
         }
 
         itemstack = this.player.inventory.getItemInHand();
-        if (itemstack != null && itemstack.count == 0) {
+        if (itemstack != null && itemstack.count <= 0) { // EMC
             this.player.inventory.items[this.player.inventory.itemInHandIndex] = null;
             itemstack = null;
         }
diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index c963fac..0488dbf 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -339,7 +339,7 @@ public class PlayerInteractManager {
                 }
             }
 
-            if (itemstack1.count == 0) {
+            if (itemstack1.count <= 0) { // EMC
                 entityhuman.inventory.items[entityhuman.inventory.itemInHandIndex] = null;
             }
 
diff --git a/src/main/java/net/minecraft/server/PlayerInventory.java b/src/main/java/net/minecraft/server/PlayerInventory.java
index c9cc47a..9328ce4 100644
--- a/src/main/java/net/minecraft/server/PlayerInventory.java
+++ b/src/main/java/net/minecraft/server/PlayerInventory.java
@@ -486,7 +486,7 @@ public class PlayerInventory implements IInventory {
 
     public ItemStack getCarried() {
         // CraftBukkit start
-        if (this.g != null && this.g.count == 0) {
+        if (this.g != null && this.g.count <= 0) { // EMC
             this.setCarried(null);
         }
         // CraftBukkit end
diff --git a/src/main/java/net/minecraft/server/Slot.java b/src/main/java/net/minecraft/server/Slot.java
index 2371c24..b5aacf0 100644
--- a/src/main/java/net/minecraft/server/Slot.java
+++ b/src/main/java/net/minecraft/server/Slot.java
@@ -48,6 +48,7 @@ public class Slot {
     }
 
     public void set(ItemStack itemstack) {
+        if (itemstack != null && itemstack.count < 0) itemstack.count = 0; // EMC
         this.inventory.setItem(this.index, itemstack);
         this.f();
     }
diff --git a/src/main/java/net/minecraft/server/TileEntityDispenser.java b/src/main/java/net/minecraft/server/TileEntityDispenser.java
index ab91462..9eb11d4 100644
--- a/src/main/java/net/minecraft/server/TileEntityDispenser.java
+++ b/src/main/java/net/minecraft/server/TileEntityDispenser.java
@@ -100,6 +100,7 @@ public class TileEntityDispenser extends TileEntity implements IInventory {
 
     public void setItem(int i, ItemStack itemstack) {
         this.items[i] = itemstack;
+        if (itemstack != null && itemstack.count < 0) itemstack.count = 0; // EMC
         if (itemstack != null && itemstack.count > this.getMaxStackSize()) {
             itemstack.count = this.getMaxStackSize();
         }
-- 
1.8.3.2

